### PR TITLE
Suggest adding a missing zero to a floating point number

### LINKED ISSUE #98836: Suggest adding zero to errornous floating point number with dot and exponent but no post-dot digit
When writing floating point numbers, some people try to include the dot and and exponent notation. Rust sees this as a field access on an integral value. It should suggest adding a `0` after the dot.

Given the following code: [Playpen](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a4120491cf37cccc22a6b19e653c1491)

```rust
fn main() {
    let n = 2.e1;
}
```

The current output is:

```
Compiling playground v0.0.1 (/playground)
error[[E0610]](https://doc.rust-lang.org/stable/error-index.html#E0610): `{integer}` is a primitive type and therefore doesn't have fields
 --> src/main.rs:2:15
  |
2 |     let n = 2.e1;
  |               ^^

For more information about this error, try `rustc --explain E0610`.
error: could not compile `playground` due to previous error
```

Ideally the output should suggest adding a zero, like as follows:

```
Compiling playground v0.0.1 (/playground)
error[[E0610]](https://doc.rust-lang.org/stable/error-index.html#E0610): `{integer}` is a primitive type and therefore doesn't have fields
 --> src/main.rs:2:15
  |
2 |     let n = 2.e1;
  |               ^^

For more information about this error, try `rustc --explain E0610`.
error: could not compile `playground` due to previous error

help: If the number is meant to be a floating point number, add a `0` after the period.
 --> src/main.rs:2:15
  |
2 |     let n = 2.0e1;
  |                 ^
```

It might also be good to squelch the E0610 message altogether too.

### PR DESCRIPTION
**Pull request by TaKO8Ki**
fixes #98836

### CODE REVIEW

**Comment by TaKO8Ki** [comment]
I will also implement this suggestion for the following case.

```rust
fn main() {
    3.e1f32;
}
```

File: compiler/rustc_typeck/src/check/expr.rs [hunk 1]
```diff
@@ -2162,14 +2162,31 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+            let mut err = type_error_struct!(
                 self.tcx().sess,
                 field.span,
                 expr_t,
                 E0610,
                 "`{expr_t}` is a primitive type and therefore doesn't have fields",
-            )
-            .emit();
+            );
+            if expr_t.is_integral()
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/check/expr.rs:2173]: Is it possible for you to check that  the type we're dereferencing with `.` is a numerical literal? Specifically, I don't want it to fix:

```
let x = 1i32;
x.e10;
```

Comment (compiler-errors) [reply compiler/rustc_typeck/src/check/expr.rs:2173]: Or if it already accounts for this, can you add a unit test for this case?

Comment (compiler-errors) [reply compiler/rustc_typeck/src/check/expr.rs:2173]: To elaborate on this, I would probably just `span_to_snippet` and check that it parses as a valid number. You should also add a check for 

```
macro_rules! num { () => { 1 } }

fn main() {
    num!().e10;
}
```

**Comment by compiler-errors** [comment]
Thanks for the thorough test cases, when I am at my computer I will double check the implementation but it looks great.

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
r=me after CI turns green

@bors delegate+

**Pull request merged**